### DISCUSSION TITLE
Does the Kitty keyboard protocol accommodate sending `Shift`+`Space`?

### DISCUSSION THREAD

**Discussion by tkf144** [kovidgoyal/kitty#5342] [Q&A] [answered]
Edit: on further investigation, I may have misunderstood/overestimated how much of the protocol Neovim has adopted.

Hi,

I like to use `space` for entering insert mode in Vim (Neovim 0.7.2) and `shift`+`space` for exiting insert mode. Until now, I've used mechanisms like Kitty's `map` (i.e. `map shift+space send_text all \x1b`) to treat `shift`+`space` as `escape`, because `shift`+`space` is not a supported/recognised combination due to legacy terminal reasons I've long forgotten.

Recently I was happy to learn of [Kitty's keyboard protocol](https://sw.kovidgoyal.net/kitty/keyboard-protocol/) when I saw Neovim had added support. Assuming I've not misunderstood, this new protocol allows terminal programs to opt in to more accurate/complete key press information, e.g. `shift`+`space` would be sent through to Neovim instead of simply `space`.

However I'm unable to have Kitty or Neovim recognise `shift`+`space`; I simply get `' '` (quotes mine). Combinations such as `control`+`shift`+`space` and `control`+`alt`+`space` work OK. To be explicit, I hope to write a line such as `imap <S-Space> <Esc>`.

Is this possible?

Below is some basic testing I've tried. Apologies for the verbosity.

1. Start a 'configless' Kitty instance using `echo "" | kitty --config=-`
1. Run `kitty +kitten show_key -m kitty`
1. Press `shift`, release `shift`
1. Press `space`, release `space`
1. Press `shift`, press `space`, release `space`, release `shift`

![image](https://user-images.githubusercontent.com/17581159/182087122-79d05c67-02a1-4b85-a85c-e9d15b1380aa.png)

1. From the 'configless' Kitty instance established above
1. Start a 'configless' Neovim instance using `nvim --clean`
1. Type `:map ` (this is insignificant; it's just how I arrived at the screenshot below)
1. Press `control`+`v` to have Neovim output the subsequent key press
1. Press `control`+`shift`+`space`
1. Press `space` to separate the keypress output
1. Press `control`+`v` to have Neovim output the subsequent key press
1. Press `control`+`alt`+`space`
1. Press `space` to separate the keypress output, and then `'` to delimit the output of the following keypress
1. Press `control`+`v` to have Neovim output the subsequent key press
1. Press `shift`+`space`
1. Press `'` to delimit the output of the previous keypress

![image](https://user-images.githubusercontent.com/17581159/182087154-8be0c78f-0554-46eb-995f-699b74fb0676.png)

Thanks!

**Comment by kovidgoyal** [comment, marked as answer]
yes it does as your output from the show_key kitten demonstrates. 

shift+space is CSI 32 ; 2 ; 32 
space without shift is CSI 32 ;  ; 32 u

Why neovim doesnt support it, you will need to ask there. IIRC neovim's current keyboard efforts are fairly buggy you should try running neovim nightly or search its issue tracker.

**Reply by tkf144** [reply]
Thanks @kovidgoyal, appreciate the quick response. I will try a nightly Neovim.

**Reply by page-down** [reply]
I asked this question here.
https://github.com/neovim/neovim/pull/19180#issuecomment-1214276379

So when Neovim is currently only enabled for `0b1 (Disambiguate escape codes)`, the protocol documentation is clear that `Shift+Space` will not send the `CSI u` escape sequence.

https://sw.kovidgoyal.net/kitty/keyboard-protocol/#disambiguate
> ... all key events that do not generate text ...

So you need to wait for Neovim to support the full kitty keyboard protocol. Or write kitten to check if the current foreground process is nvim and send another uncommon shortcut for `Shift+Space`.

**Reply by primeapple** [reply]
I have some troubles understanding this. I cannot map `C-/` and `C--` in Neovim. Is this also because of the same bug/missing implementation?

**Reply by page-down** [reply]
@primeapple
These two key combinations work fine, and I'm using neovim nightly under macOS.
It is likely that you are using an older version of neovim or some other environment configuration issue. There is no bug in kitty or the protocol itself.
You need to ask in the neovim project community to find out more.

In any case, I don't believe your issue is relevant to what was discussed in this thread.

**Comment by chdiza** [comment]
Incidentally, I find that regular Vim won't let me map `<S-Space>` while in kitty; or rather, Vim lets me map it, but pressing Shift-Space doesn't trigger the map.

Vim now claims to support the kitty keyboard protocol, but I'm not sure if they *actually* support it.

**Comment by kovidgoyal** [comment]
You'd need to report it there.

**Reply by chdiza** [reply]
Of course :) 

I just thought perhaps the OP would like to know that this isn't *only* a neovim thing.

**Comment by itachi-19** [comment]
@tkf144 in case you were unable to do it, here's what worked for me:

In kitty.conf:
map shift+space send_text all \x1b[32;2u

Then shift + space can be mapped using <S-Space>
